### PR TITLE
Bugfixing Data Foundations

### PR DESCRIPTION
**Pull request by javiergp**
- Fixed SA permissions
- Usage of impersonation to avoid SA private key export
- Fixed required API enablement
- Added FW rules required by dataflow

### CODE REVIEW

File: data-solutions/data-platform-foundations/01-environment/main.tf [hunk 1]
```diff
@@ -31,8 +31,9 @@ module "project-datamart" {
     "storage.googleapis.com",
     "storage-component.googleapis.com",
   ]
-  iam = {
-    "roles/editor" = [module.sa-services-main.iam_email]
+
+  iam_additive = {
```
Comment (ludoo) [inline data-solutions/data-platform-foundations/01-environment/main.tf:35]: did you try running `apply` with an empty state? `iam_additive` is not safe with dynamic values.

File: data-solutions/data-platform-foundations/02-resources/README.md [hunk 2]
```diff
@@ -38,15 +38,15 @@ project_ids = {
...
-- Get a key for the service account created in the environment stage:
-  - Go into services project
-  - Go into IAM page
-  - Go into the service account section
-  - Creaet a new key for the service account created in previeous step (**service_account**)
-  - Download the json key into the current folder
-- make sure you have the right authentication setup: `export GOOGLE_APPLICATION_CREDENTIALS=PATH_TO_SERVICE_ACCOUT_KEY.json`
-- run `terraform init` and `terraform apply`
 
+- Set up the gcloud environment to impersonate the **service_account**: `gcloud config set auth/impersonate_service_account **service_account**`
```
Comment (ludoo) [inline data-solutions/data-platform-foundations/02-resources/README.md:42]: An easier way would be to precreate backend and provider in the previous step, setting up impersionation.

File: data-solutions/data-platform-foundations/02-resources/main.tf [hunk 3]
```diff
@@ -147,6 +166,31 @@ module "vpc-transformation" {
...
+  ssh_source_ranges     = []
+
+  custom_rules = {
+    iap-svc = {
+      description          = "Dataflow service."
+      direction            = "INGRESS"
+      action               = "allow"
+      sources              = ["dataflow"]
+      targets              = ["dataflow"]
+      ranges               = []
```
Comment (ludoo) [inline data-solutions/data-platform-foundations/02-resources/main.tf:186]: if this is for IAP, it should use the IAP source ranges

**Review by ludoo — CHANGES_REQUESTED** [review]
Thanks for the PR, I dropped a few comments.

File: data-solutions/data-platform-foundations/02-resources/README.md [hunk 4]
```diff
@@ -38,15 +38,15 @@ project_ids = {
...
+
+- Generate an access token and launch terraform:
+```bash
+export GOOGLE_OAUTH_ACCESS_TOKEN=$(gcloud auth print-access-token)
```
Comment (juliocc) [inline data-solutions/data-platform-foundations/02-resources/README.md:46]: the GCP provider supports this natively now through the [impersonate_service_account](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/provider_reference#impersonate_service_account) option.

File: data-solutions/data-platform-foundations/02-resources/main.tf [hunk 5]
```diff
@@ -25,21 +25,30 @@ module "datamart-sa" {
...
   }
+  iam = var.admins != null ?  { "roles/iam.serviceAccountTokenCreator" =  concat(var.admins) } : {}
 }
 
 module "dwh-sa" {
   source     = "../../../modules/iam-service-account"
   project_id = var.project_ids.dwh
   name       = var.service_account_names.dwh
+  
+   iam_project_roles = {
```
Comment (juliocc) [inline data-solutions/data-platform-foundations/02-resources/main.tf:36]: I think the linter will fail here. Make sure to run all terraform files through `terarform fmt`

File: data-solutions/data-platform-foundations/01-environment/variables.tf [hunk 6]
```diff
@@ -74,3 +74,8 @@ variable "service_perimeter_standard" {
   type        = string
   default     = null
 }
+
+variable "admins" {
```
Comment (juliocc) [inline data-solutions/data-platform-foundations/01-environment/variables.tf:78]: keep variables sorted alphabetically

File: data-solutions/data-platform-foundations/02-resources/README.md [hunk 7]
```diff
@@ -26,7 +26,7 @@ In the previous step, we created the environment (projects and service account)
 
 To create the resources, copy the output of the environment step (**project_ids**) and paste it into the `terraform.tvars`:
 
-- Specify your variables in a `terraform.tvars`, you can use the ouptu from the environment stage
+- Specify your variables in a `terraform.tvars`, you can use the output from the environment stage
```
Comment (juliocc) [inline data-solutions/data-platform-foundations/02-resources/README.md:29]: Recommendation here: generate the tfvars file content as an output from the previous stage.

**Review by juliocc — CHANGES_REQUESTED** [review]

File: data-solutions/data-platform-foundations/01-environment/main.tf (hunk 1 above)
Comment (juliocc) [reply data-solutions/data-platform-foundations/01-environment/main.tf:35]: True, but `iam_email` is computed [statically](https://github.com/terraform-google-modules/cloud-foundation-fabric/blob/master/modules/iam-service-account/outputs.tf#L32) now. This particular usage should be fine

File: data-solutions/data-platform-foundations/02-resources/main.tf (hunk 5 above)
Comment (ludoo) [reply data-solutions/data-platform-foundations/02-resources/main.tf:36]: :+1: you can set vscode to format on save, and forget about formatting in the future

File: data-solutions/data-platform-foundations/01-environment/main.tf (hunk 1 above)
Comment (javiergp) [reply data-solutions/data-platform-foundations/01-environment/main.tf:35]: The main issue with authoritative binding for roles/editor was that it removed the binding for some Google owned SA (i.e. Google APIs Service Agent). This leaded to issues when deploying Composer, for instance. terraform apply works OK in a from the scratch deployment

File: data-solutions/data-platform-foundations/02-resources/main.tf (hunk 5 above)
Comment (javiergp) [reply data-solutions/data-platform-foundations/02-resources/main.tf]: fixed

File: data-solutions/data-platform-foundations/01-environment/variables.tf (hunk 6 above)
Comment (javiergp) [reply data-solutions/data-platform-foundations/01-environment/variables.tf]: fixed

File: data-solutions/data-platform-foundations/02-resources/README.md (hunk 7 above)
Comment (javiergp) [reply data-solutions/data-platform-foundations/02-resources/README.md:29]: will fix in next release

File: data-solutions/data-platform-foundations/02-resources/README.md (hunk 2 above)
Comment (javiergp) [reply data-solutions/data-platform-foundations/02-resources/README.md]: fixed

File: data-solutions/data-platform-foundations/02-resources/main.tf (hunk 3 above)
Comment (javiergp) [reply data-solutions/data-platform-foundations/02-resources/main.tf:186]: Nope, this is only for dataflow

File: data-solutions/data-platform-foundations/01-environment/main.tf [hunk 8]
```diff
@@ -153,6 +157,6 @@ module "sa-services-main" {
   source     = "../../../modules/iam-service-account"
   project_id = module.project-services.project_id
   name       = var.service_account_names.main
-  iam        = var.admins != null ?  { "roles/iam.serviceAccountTokenCreator" =  concat(var.admins) } : {}
+  iam        = var.admins != null ? { "roles/iam.serviceAccountTokenCreator" = concat(var.admins) } : {}
```
[the remark this answers is not in the document]
Comment (javiergp) [reply data-solutions/data-platform-foundations/01-environment/main.tf]: Removed note needed concat()

File: data-solutions/data-platform-foundations/02-resources/README.md [hunk 9]
```diff
@@ -38,15 +38,14 @@ project_ids = {
...
-  - Creaet a new key for the service account created in previeous step (**service_account**)
-  - Download the json key into the current folder
-- make sure you have the right authentication setup: `export GOOGLE_APPLICATION_CREDENTIALS=PATH_TO_SERVICE_ACCOUT_KEY.json`
-- run `terraform init` and `terraform apply`
 
+- The providers.tf file has been configured to impersonate the **main** service account
+
+- To launch terraform:
+```bash
+terraform plan
```
Comment (ludoo) [inline data-solutions/data-platform-foundations/02-resources/README.md:46]: nit: `plan` is redundant

File: data-solutions/data-platform-foundations/02-resources/main.tf [hunk 10]
```diff
@@ -25,21 +25,30 @@ module "datamart-sa" {
...
 
 module "landing-sa" {
   source     = "../../../modules/iam-service-account"
   project_id = var.project_ids.landing
   name       = var.service_account_names.landing
   iam_project_roles = {
-    "${var.project_ids.landing}" = ["roles/pubsub.publisher"]
+    "${var.project_ids.landing}" = [
+      "roles/pubsub.publisher",
+    "roles/storage.objectCreator"]
```
Comment (ludoo) [inline data-solutions/data-platform-foundations/02-resources/main.tf:49]: nit: add newline before closing brace

**Review by ludoo — APPROVED** [review]
couple of minor nits

**Review by juliocc — APPROVED** [review]

**Pull request merged**